---
title: "YZhu Ribotag RNAseq re-analysis"
author: "James Choi"
date: "Last compiled:`r Sys.Date()`"
output: 
  html_document
editor_options:
  chunk_output_type: 
    console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align='center', 
                      tidy.opts=list(width.cutoff=80), tidy=TRUE,
                      results='hold')
```


## Background

See Yunjiao Zhu's paper: [Macrophage Transcriptional Profile Identifies Lipid Catabolic Pathways That Can Be Therapeutically Targeted after Spinal Cord Injury](https://doi.org/10.1523/JNEUROSCI.2751-16.2017).

The goal of this analysis is to replicate YZ's findings. YZ's downstream 
analysis used FPKMs, which is not compatible with the count matrices produced
by Christine Ryan's in vitro foam cell RNAseq data and my scRNAseq data. To 
produce a count matrix for YZ's data, I re-ran read alignment and gene counting
using tools made available through [Galaxy](https://usegalaxy.org/). 

YZhu's Ribotag RNAseq data is paired-end, unstranded reads prepared using
Illumina's ScriptSeq v2 RNA-Seq Library Preparation Kit but sequenced using
Illumina's TruSeq sequencing kit (as of 2021-10-19, I'm not entirely sure but
think that the TruSeq kit is distinct from the ScriptSeq kit. TruSeq adapters
and sequences were used for fastq file manipulation).

In brief, the following was performed:

1. Upload fastq files via SRR_Acc_list.txt
    + SRR3945463    
    + SRR3945464
    + SRR3945465
    + SRR3945466
    + SRR3945467
    + SRR3945468
2. FastQC on raw fastq files (see: FastQC_Reports directory)
    + Presence of Adapter sequences in reads led to decision to trim reads.
3. Cutadapt on raw fastq files
    + R1 3' adapter sequence: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC (Illumina's TruSeq Index Adapter; from Galaxy's description box; see "TUCF_Understanding_Illumina_TruSeq_Adapters.pdf")
    + R2 3' adapter sequence: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT (Illumina's TruSeq Universal Adapter)
4. FastQC on trimmed reads (post-Cutadapt fastq files)
5. RNA STAR on trimmed reads to produce bam files
    + Built-in reference genome: index (see Galaxy RNA STAR page's options)
    + Reference genome: mm10
    + Gene model file for splice junctions: gencode.vM27.annotation.gtf.gz (from: http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz)
    + Length of genomic sequence around annotated junctions: 100
    + MAPQ value for unique mappers: 60
6. Infer Experiment on bam files 
    + Since strandedness was not known a priori, I tried using RSeQC's infer_experiement.py. Output showed 65%-35% strandedness, suggesting reads were unstranded. 
7. featureCounts on bam files
    + Do not use gtf file from STAR mapping for featureCounts because it leads to
  ~60% of reads unassigned_noFeature. Not sure why this happens.
    + Specify strand information	Unstranded
    + Gene annotation file	builtin
    + Select built-in genome	mm10
    + Output format	Gene-ID "\t" read-count (MultiQC/DESeq2/edgeR/limma-voom compatible)
    + Count fragments instead of reads	-p
    + Only allow fragments with both reads aligned	True
    + On feature level	False
    + Allow reads to map to multiple features		(default, dont allow
    multimapping)


## Setup

```{r libraries}
require('ggplot2')
require('dplyr')
require('patchwork')
require('DESeq2')
require('edgeR')
```

```{r directories}
if (!grepl('scripts', getwd())) {setwd('scripts')}
data_path <- '../data/'
results_out <- '../results/Zhu_RNAseq_replication/'
dir.create(path = results_out)
```

```{r utils}
my_theme <- theme_bw() +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 12),
        strip.text = element_text(size = 12),
        plot.title = element_text(size = 14),
        plot.subtitle = element_text(size = 12))
```


## Data import, cleaning

I import YZhu's FPKM data, which is available from GEO using the [Series Accession number GSE84737](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84737). I imported the [GSE84737_All_genes_unfiltered_DE.txt.gz file](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE84737&format=file&file=GSE84737%5FAll%5Fgenes%5Funfiltered%5FDE%2Etxt%2Egz).


### Import data

```{r fpkm_import}
fpkm <- download.file(
  url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE84737&format=file&file=GSE84737%5FAll%5Fgenes%5Funfiltered%5FDE%2Etxt%2Egz',
  destfile = '../data/GSE84737_All_genes_unfiltered_DE.txt.gz'
)
fpkm <- read.table(
  file = gzfile('../data/GSE84737_All_genes_unfiltered_DE.txt.gz'),
  header = TRUE,
  sep = '\t'
)
# closeAllConnections()
```


```{r rerun_import}
rerun_files <- list.files(
  path = '../results/Zhu_RNAseq_Galaxy/featureCounts_counts/',
  full.names = TRUE
)
```


### Wrangling

Separate each dataset into gene-related information and the actual expression
counts.


```{r fpkm_wrangling}
unique_genes <- which(!duplicated(fpkm$official_gene_name))
fpkm_gene_info <- fpkm[unique_genes, c('official_gene_name','locus','q_value')]
fpkm_counts <- fpkm[c(3:8)]
fpkm_counts <- fpkm_counts[unique_genes,]
colnames(fpkm_counts) <- c('D3_rep1','D3_rep2','D3_rep3',
                           'D7_rep1','D7_rep2','D7_rep3')
rownames(fpkm_counts) <- fpkm_gene_info$official_gene_name
```


```{r rerun_wrangling}
rerun_names <- strsplit(x = rerun_files,  split = '/')
rerun_names <- sapply(rerun_names, `[`, 5)
rerun_names <- strsplit(x = rerun_names, split = ' ')
rerun_names <- sapply(rerun_names, `[`, 3)
rerun_names <- gsub(
  pattern = '\\.tabular',
  replacement = '',
  x = rerun_names
)
rerun_names <- paste(
  c('D3_rep1','D3_rep2','D3_rep3','D7_rep1','D7_rep2','D7_rep3'),
  rerun_names,
  sep = '_'
)
rerun <- vector(mode = 'list', length = length(rerun_files))
names(rerun) <- rerun_names
for (i in 1:length(rerun_files)) {
  rerun[[i]] <- read.table(
    file = rerun_files[i],
    sep = '\t',
    header = TRUE
  )
  colnames(rerun[[i]]) <- c('Gene_id', rerun_names[i])
}

# Import converted gene IDs
gene_info_files <- list.files(
  path = '../results/Zhu_RNAseq_Galaxy/featureCounts_gene_ID_convert/',
  full.names = TRUE
)
gene_info <- vector(mode = 'list', length = length(gene_info_files))
for (i in 1:length(gene_info)) {
  gene_info[[i]] <- read.table(
    file = gene_info_files[i],
    sep = '\t',
    header = TRUE
  )
}
# identical(gene_info[[1]], gene_info[[4]])
gene_info <- gene_info[[1]]

# Clean up gene IDs. Decimal places indicate gene ID versions, can simply be cut
for (i in 1:length(rerun)) {
  rerun[[i]]$Ensembl <- plyr::mapvalues(
    x = rerun[[i]]$Gene_id,
    from = gene_info$ENTREZID,
    to = gene_info$ENSEMBL,
    warn_missing = FALSE
  )
  rerun[[i]]$Gene_name <- plyr::mapvalues(
    x = rerun[[i]]$Gene_id,
    from = gene_info$ENTREZID,
    to = gene_info$SYMBOL,
    warn_missing = FALSE
  )
  rerun[[i]] <- rerun[[i]][c(1,3,4,2)]
}
rerun <- Reduce(f = left_join, x = rerun)
write.table(rerun, file = '../data/YZhu_Ribotag_Reanalysis_countMatrix.txt',
            sep = '\t', row.names = FALSE, quote = FALSE)

# Filter genes by availability of unique Ensembl ID
rerun_filtered <- rerun %>% 
  filter(!is.na(Gene_name)) %>% 
  filter(!duplicated(Gene_name))

# counts and gene_info
rerun_gene_info <- rerun_filtered[1:3]
rerun_counts <- rerun_filtered[4:ncol(rerun_filtered)]
rownames(rerun_counts) <- rerun_filtered$Gene_name
colnames(rerun_counts) <- c('D3_rep1','D3_rep2','D3_rep3',
                            'D7_rep1','D7_rep2','D7_rep3')
```


## Differentially expressed genes

### YZhu FPKM data

This is largely as-is.

```{r fpkm_deg}
fpkm_deg_n <- fpkm %>% 
  filter(q_value < 0.01) %>%
  mutate('dir' = ifelse(log2.fold_change. > 0, 'd7', 'd3')) %>% 
  group_by(dir) %>% 
  summarise(ngenes = n_distinct(official_gene_name))
min_qval <- ceiling(max(-log10(fpkm$q_value)))
max_fc <- max(abs(fpkm$log2.fold_change.[!is.infinite(fpkm$log2.fold_change.)]))
interest_genes <- c('Cd5l','Lpl','Ccl2','Mmp12','Igf1','Cxcl3','Plac8','Apoe')
tmp <- fpkm %>% 
  mutate(log_qval = -log10(q_value)) %>% 
  mutate(log_qval = ifelse(is.infinite(log_qval), min_qval, log_qval)) %>% 
  mutate(significant = ifelse(q_value < 0.01, 'yes', 'no'))
fpkm_deg_volcano <- tmp %>% 
  ggplot(mapping = aes(x = log2.fold_change., y = log_qval)) +
  geom_point(mapping = aes(color = significant), size = 1) +
  ggrepel::geom_label_repel(data = tmp[tmp$official_gene_name %in% interest_genes,],
             mapping = aes(label = official_gene_name)) +
  scale_color_manual(values = c('yes' = 'black', 'no' = 'grey80')) +
  scale_x_continuous(limits = c(-max_fc, max_fc),
                     breaks = seq(-10, 10, 2)) +
  xlab(label = 'log2(fold-change)') +
  ylab(label = '-log10(q-value)') +
  labs(title = 'FPKM DEG (original study)',
       subtitle = 'logFC > 0 indicates higher in 7dpi') +
  my_theme
```


```{r fpkm_maplot}
tmp <- data.frame(
  logmean = log10(rowMeans(fpkm_counts)),
  mean_3d = rowMeans(fpkm_counts[grepl('D3', colnames(fpkm_counts))]),
  mean_7d = rowMeans(fpkm_counts[grepl('D7', colnames(fpkm_counts))]),
  q_value = fpkm_gene_info$q_value
) %>% 
  mutate(GeneName = rownames(.)) %>% 
  mutate(log2fc = log2(mean_7d/mean_3d)) %>% 
  mutate(significant = ifelse(q_value < 0.01, 'yes', 'no')) %>% 
  arrange(significant)
fpkm_ma <- tmp %>% 
  ggplot(mapping = aes(x = logmean, y = log2fc)) +
  geom_point(mapping = aes(color = significant),
             size = 1) +
  ggrepel::geom_label_repel(
    data = tmp[interest_genes,],
    mapping = aes(label = GeneName)) +
  scale_color_manual(values = c('yes' = 'black', 'no' = 'grey80')) +
  scale_y_continuous(limits = c(-max_fc, max_fc)) +
  ylab(label = 'log2(fold-change)') +
  xlab(label = 'log10(mean FPKM)') +
  my_theme
```


### Re-run count data

For the rerun data, I go through the DESeq2 pipeline.

```{r rerun_library_size, fig.height=4, fig.width=4}
conditions <- factor(sapply(strsplit(colnames(rerun_counts), '_'), `[`, 1))
sample_stats <- data.frame(
  lib.size = colSums(rerun_counts),
  condition = conditions
)
boxplot(formula = lib.size ~ condition, 
        data = sample_stats,
        main = 're-run counts library size',
        ylab = 'sum reads per sample')
points(x = sample_stats$condition[1:3], y = sample_stats$lib.size[1:3],
       cex = 1.3)
points(x = sample_stats$condition[4:6], y = sample_stats$lib.size[4:6],
       cex = 1.3)
```


```{r rerun_mds, fig.height=4.5, fig.width=4.5}
dge <- DGEList(counts = rerun_counts, group = conditions)
design <- model.matrix( ~ group + 0, data = dge$samples)
colnames(design) = gsub("group","",colnames(design))
plotMDS(dge, top=500,  
        col=c(rep("black",3), rep("red",3)), 
        main = 'MDS of log-transformed counts')
```


Plotting mean-variance curve tells us something about how the statistical tests
will define differential expression.

```{r rerun_meanvar, fig.height=4.5, fig.width=5}
log2p <- function(x) log2(x + 1)
gene_stats <- data.frame(
  mean = apply(X = log2p(rerun_counts), MARGIN = 1, FUN = mean),
  var = apply(X = log2p(rerun_counts), MARGIN = 1, FUN = var)
)
gene_stats <- gene_stats[order(gene_stats$mean),]
meanvar_fit <- loess(formula = var ~ mean, data = gene_stats, span = 0.3)
meanvar_smooth <- predict(meanvar_fit)

plot(x = gene_stats$mean,
     y = gene_stats$var,
     xlab = 'Average log2(count+1)',
     ylab = 'Variance log2(counts+1)',
     main = 'rerun variance-stabilization',
     sub = 'gene mean vs variance modeling')
lines(predict(meanvar_fit), x = gene_stats$mean, col = 'red', lwd = 2)
```


```{r deseq2}
dds <- DESeqDataSetFromMatrix(countData = as.matrix(rerun_counts),
                              colData = sample_stats,
                              design = ~ condition)
mcols(dds) <- cbind(mcols(dds), rerun_gene_info)
keep <- rowSums(counts(dds)) >= 6
dds <- dds[keep,]
# Set reference such that logFC > 0 denotes higher in 7d
dds$condition <- relevel(x = dds$condition, ref = 'D3')
dds <- estimateSizeFactors(dds)
dds <- DESeq(object = dds)
rerun_result <- lfcShrink(dds, coef = c('condition_D7_vs_D3'))
rerun_result <- as.data.frame(rerun_result)
rerun_result$GeneName <- rownames(rerun_result)
```

```{r deseq_plots}
rerun_deg_n <- rerun_result %>% 
  filter(padj < 0.01) %>%
  mutate('dir' = ifelse(log2FoldChange > 0, 'd7', 'd3')) %>% 
  group_by(dir) %>% 
  summarise(ngenes = n_distinct(GeneName))
min_qval <- ceiling(-log10(min(rerun_result$padj, na.rm = TRUE)))
max_fc <- ceiling(max(abs(rerun_result$log2FoldChange))*10)/10
interest_genes <- c('Cd5l','Lpl','Ccl2','Mmp12','Igf1','Cxcl3','Plac8','Apoe')
tmp <- rerun_result %>% 
  mutate(log_qval = -log10(padj)) %>% 
  mutate(significant = ifelse(padj < 0.01, 'yes', 'no'))
rerun_deg_volcano <- tmp %>% 
  ggplot(mapping = aes(x = log2FoldChange, y = log_qval)) +
  geom_point(mapping = aes(color = significant), size = 1) +
  ggrepel::geom_label_repel(data = tmp[tmp$GeneName %in% interest_genes,],
             mapping = aes(label = GeneName)) +
  scale_color_manual(values = c('yes' = 'black', 'no' = 'grey80'),
                     na.value = 'grey80') +
  scale_x_continuous(limits = c(-max_fc, max_fc),
                     breaks = seq(-10, 10, 2)) +
  xlab(label = 'log2(fold-change)') +
  ylab(label = '-log10(q-value)') +
  labs(title = 'Re-run DEG (replication study)',
       subtitle = 'logFC > 0 indicates higher in 7dpi') +
  my_theme
```

```{r rerun_maplot}
tmp <- rerun_result %>% 
  mutate(logmean = log10(baseMean)) %>% 
  mutate(significant = ifelse(padj < 0.01, 'yes', 'no')) %>% 
  arrange(significant)
rerun_ma <- tmp %>% 
  ggplot(mapping = aes(x = logmean, y = log2FoldChange)) +
  geom_point(mapping = aes(color = significant),
             size = 1) +
  ggrepel::geom_label_repel(data = tmp[tmp$GeneName %in% interest_genes,],
             mapping = aes(label = GeneName)) +
  scale_color_manual(values = c('yes' = 'black', 'no' = 'grey80'),
                     na.value = 'grey80') +
  scale_y_continuous(limits = c(-max_fc, max_fc)) +
  ylab(label = 'log2(shrinked fold-change)') +
  xlab(label = 'log10(mean count)') +
  my_theme
```


## FPKM vs featureCounts comparison

```{r deg_summary, fig.height=8, fig.width=10}
volcano <- fpkm_deg_volcano | rerun_deg_volcano
maplot <- fpkm_ma | rerun_ma
summary_fig <- (volcano / maplot) + plot_layout(guides = 'collect')
summary_fig
```

```{r deg_venn, fig.height=4, fig.width=5}
fpkm_deg <- split(x = fpkm$official_gene_name[fpkm$q_value < 0.01], 
                  f = fpkm$log2.fold_change.[fpkm$q_value < 0.01] > 0)
names(fpkm_deg) <- c('D7','D3')
rerun_deg <- split(x = rerun_result$GeneName[rerun_result$padj < 0.01],
                   f = rerun_result$log2FoldChange[rerun_result$padj < 0.01] >0)
names(rerun_deg) <- c('D7','D3')
d7venn <- eulerr::euler(
  combinations = list('original study' = unique(fpkm_deg$D7),
                      'replication study' = unique(rerun_deg$D7)
  )
)
d3venn <- eulerr::euler(
  combinations = list('original study' = unique(fpkm_deg$D3),
                      'replication study' = unique(rerun_deg$D3)
  )
)
plot(d7venn, main = 'Genes higher at 7dpi', quantities = TRUE)
plot(d3venn, main = 'Genes higher at 3dpi', quantities = TRUE)
```


```{r}
sessionInfo()
```

